### PR TITLE
Bring back BOSH session connectionTimeout

### PR DESCRIPTION
**Pull request by sonnyp**
During the BOSH refactor we forgot about connectionTimeout, this PR brings it back.

connectionTimeout removes the session if there wasn't any request after X seconds since the last one ended

### CODE REVIEW

File: lib/bosh/session.js [hunk 1]
```diff
@@ -250,6 +251,18 @@ BOSHServerSession.prototype.workOutQueue = function() {
...
+    this.maySetConnectionTimeout()
+}
+
+BOSHServerSession.prototype.maySetConnectionTimeout = function() {
+    if (!this.connectionTimeout && this.outQueue.length < 1) {
+        var self = this;
+        this.connectionTimeout = setTimeout(function() {
+            self.emit('error', new Error('Session timeout'))
+            self.emit('close')
+        }, 3 * 1000)
```
Comment (saimonmoore) [inline lib/bosh/session.js]: maybe extract to a constant?

**Comment by saimonmoore** [comment]
:green_apple: && :postal_horn: :+1:

File: lib/bosh/session.js [hunk 2]
```diff
@@ -65,6 +65,7 @@ function BOSHServerSession(opts) {
 
     this.inQueue[opts.bodyEl.attrs.rid] = opts
     process.nextTick(this.workInQueue.bind(this))
+    this.maySetConnectionTimeout();
```
Comment (chris-rock) [inline lib/bosh/session.js]: can we rename this to setConnectionTimeout?

File: lib/bosh/session.js [hunk 3]
```diff
@@ -250,6 +251,18 @@ BOSHServerSession.prototype.workOutQueue = function() {
...
+    this.maySetConnectionTimeout()
+}
+
+BOSHServerSession.prototype.maySetConnectionTimeout = function() {
+    if (!this.connectionTimeout && this.outQueue.length < 1) {
+        var self = this;
+        this.connectionTimeout = setTimeout(function() {
+            self.emit('error', new Error('Session timeout'))
+            self.emit('close')
+        }, 60 * 1000)
```
Comment (chris-rock) [inline lib/bosh/session.js]: please use a variable for the timeout. this enables users to configure this as needed.

File: lib/bosh/session.js (hunk 2 above)
Comment (sonnyp) [inline lib/bosh/session.js]: connectionTimeout is a bit misleading, what do you think about nextRequestTimeout ?

File: lib/bosh/session.js (hunk 3 above)
Comment (sonnyp) [inline lib/bosh/session.js]: :+1:

File: lib/bosh/session.js (hunk 2 above)
Comment (chris-rock) [inline lib/bosh/session.js]: much better. thanks

**Comment by sonnyp** [comment]
@chris-rock comments addressed

File: lib/bosh/session.js [hunk 4]
```diff
@@ -35,7 +35,14 @@ function BOSHServerSession(opts) {
     this.writable = true
     this.readable = true
 
-    // Bosh settings
+    this.options = opts.options || {};
+
+    //set default nextRequestTimeout option
+    if (!this.options.nextRequestTimeout) {
+        this.options.nextRequestTimeout = 60 * 1000
```
Comment (sonnyp) [inline lib/bosh/session.js]: this is pretty high, maybe we should set it to 10 seconds

Comment (chris-rock) [inline lib/bosh/session.js]: would like to make this configurable

Comment (lloydwatkin) [inline lib/bosh/session.js]: :+1:

Comment (sonnyp) [inline lib/bosh/session.js]: it is configurable via the server options

Options are passed from server.js to http.js to session.js

Comment (chris-rock) [inline lib/bosh/session.js]: you're right. missed it. sorry.

**Comment by sonnyp** [comment]
I made some changes, please double check.

**Comment by sonnyp** [comment]
actually wait, I found an issue

**Comment by sonnyp** [comment]
The issue is that it makes bosh server emit twice 'close' events on the session. (you can reproduce with the test case at https://github.com/node-xmpp/node-xmpp-server/issues/100

**Comment by sonnyp** [comment]
https://github.com/node-xmpp/node-xmpp-server/pull/116

**Pull request closed without merging**